File: 2. Arrays/medium/200. islands.js
// Given a 2d grid map of '1's (land) and '0's (water), count the number of islands. An island is surrounded by water and is formed by connecting adjacent lands horizontally or vertically. You may assume all four edges of the grid are all surrounded by water.

var numIslands = function(grid) {
  if (grid.length === 0) {
    return 0; // return 0 if there is no grid
  }
  let islandNum = 0;
  let rows = grid.length;
  let columns = grid[0].length; //initialize variables
  const explore = function(r, c) {
    if (r < 0 || r >= rows || c < 0 || c >= columns || grid[r][c] === "0") {
      return;
    }
    grid[r][c] = "0";
    explore(r + 1, c);
    explore(r - 1, c);
    explore(r, c + 1);
    explore(r, c - 1);
  };

  for (let i = 0; i < rows; i++) {
    for (let j = 0; j < columns; j++) { // go through grid and call recursive function on any cell that is a "1"
      let currentCell = grid[i][j];
      if (currentCell === "1") {
        islandNum++;
        explore(i, j);
      }
    }
  }
  return islandNum;
};

File: 2. Arrays/medium/200. islands.py
class Solution:
  def __init__(self):
    self.islands = 0

  def numIslands(self, grid):
    for row in range(len(grid)):
      for col in range(len(grid[0])):
        spot = grid[row][col]
        if spot == 1:
          self.islands += 1
          self.exploreIsland(row, col, grid)
    return self.islands

  def exploreIsland(self, r, c, grid):
    if 0 <= r < len(grid) and 0 <= c < len(grid[0]) and grid[r][c] == 1:
        grid[r][c] = 0
        self.exploreIsland(r, c + 1, grid)
        self.exploreIsland(r, c - 1, grid)
        self.exploreIsland(r -1, c, grid)
        self.exploreIsland(r + 1, c, grid)

solution = Solution()

solution.numIslands([[1,1,0,0,0],[1,1,0,0,0],[0,0,1,0,0], [0,0,0,1,1]])